features/FxA-45-device-registration-api/FX-DESKTOP.md

```diff
@@ -0,0 +1,122 @@
+
+On startup, Firefox should attempt to register the device with the
+FxA auth server using the name and type currently stored in the
+user's Sync preferences.
+
+Firefox will send the following information to the FxA auth server to
+register the device:
+
+* deviceName (from Sync preferences)
+* deviceType (from Sync preferences)
+* pushCallback (from Push preferences)
+* pushPublicKey (from Push preferences)
+
+### Registration success
+
+The FxA auth server will respond with:
+
+* deviceID
+* deviceName
+* deviceType
+* pushCallback
+* pushPublicKey
+
+Firefox must then store the returned deviceID.
+
+### Registration failure
+
+If registration fails because of a network error, Firefox should re-attempt
+to register after a TBD interval has elapsed.
+
+If registration fails for other reasons - WHAT HERE?
+
+## Existing profile, has FxA deviceID, has FxA sessionToken
+
+This is a user who has previously signed in to FxA to
+connect to Sync and their device is already registered.
+
+This device has already sent its initial name to the FxA
+auth server, but the user may have remotely disconnected
+the device, remotely reset the account password, or remotely
+changed the account password. The FxA auth server is considered
+the canonical source of truth for device state.
+
+On startup, Firefox should fetch the current device list
+from the FxA auth server and compare internal Firefox state
+with the state returned from the auth server.
+
+### Device list fetch success
+
+The FxA auth server will respond with an array of devices with the
+following information:
+
+* deviceID
+* deviceName
+* deviceType
+* pushCallback
+* pushPublicKey
+
+If the user has remotely disconnected the device, Firefox
+should disconnect the user from Sync (and wipe their FxA sessionToken?)
+
+If the user has remotely changed or reset the account password,
+Firefox should ask the user to re-authenticate with Firefox accounts.
+
+### Device list fetch failure
+
+If device list fails because of a network error, Firefox should re-attempt
+to re-fetch the list after a TBD interval has elapsed.
+
+## Firefox receives an `fxaccounts:login` message
+
+A user has just authenticated with Firefox Accounts. The message payload
+will contain both a sessionToken.
+
+If Firefox already has an FxA deviceID, then the user re-authenticated
+for some reason (reset password, change password).
+
+XXX Will the devices API automatically re-associate the old deviceID with the
+new sessionToken when the user signs in? Is there any mechanism for the auth
+server to do that automatically?
+
+If Firefox does not have an FxA deviceID, then the device should be registered.
+See the section titled `Existing profile, no FxA deviceID, has FxA
+sessionToken`
+
+## Firefox receives an `fxaccounts:logout` message
+
+The user has just signed out of Firefox Accounts by removing the current device. The stored sessionToken and deviceID should be deleted, and the browser should disconnect from Sync.
```

Comment:
In fact I think we should be a little more explicit here and state that the act of "disconnecting from sync", no matter how it's triggered, should involve destroying the device record and session token and discarding sessionToken and deviceID.